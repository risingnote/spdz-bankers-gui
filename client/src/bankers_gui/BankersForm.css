.bankersForm {
  margin-right: 3rem;
  text-align: left;
}

.bankersForm label {
  display: block;
}

.bankersForm input {
  display: block;
  margin-bottom: 10px;
}

.diners {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 2rem;
}

.atable {
  background: lightyellow;
  border: thin solid grey;
  width: 18rem;
  height: 18rem;
  border-radius: 50%;
  position: relative;
}

.aplace {
  background: white;
  border: thin solid grey;
  width: 4rem;
  height: 4rem;
  border-radius: 10%;
  position: absolute;
  /* centre in container */
  top: 50%;
  left: 50%;
  margin: -2rem 0 0 -2rem;
  /* centre contained text */
  text-align: center;
  vertical-align: middle;
  line-height: 4rem;
}

/* Translation is 
 * (radius.sinØ, -radiuscosØ)
*/