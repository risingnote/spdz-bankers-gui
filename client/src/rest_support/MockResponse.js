/**
 * Provide a function for fetch request to determine if request was successful
 */
export default function MockResponse (status, statusText, response) {
  return new window.Response(response, {
    status: status,
    statusText: statusText,
    headers: {
      'Content-type': 'application/json'
    }
  });
};