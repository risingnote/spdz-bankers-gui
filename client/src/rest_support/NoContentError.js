// Error to represent missing data when requested from SPDZ proxy.
function NoContentError(message) {
  this.name = 'NoContentError'
  this.message = message || 'No content found.'
  this.stack = (new Error()).stack
}
NoContentError.prototype = Object.create(Error.prototype)
NoContentError.prototype.constructor = NoContentError

export default NoContentError