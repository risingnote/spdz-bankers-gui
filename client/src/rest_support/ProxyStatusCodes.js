export default Object.freeze({'Disconnected':1, 'Connected':2, 'Failure':3})