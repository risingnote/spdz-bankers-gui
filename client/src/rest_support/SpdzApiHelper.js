/**
 * Higher level functions to interact with the SPDZ api.
 */
import { consumeDataFromProxies, sendInputsToProxies } from './SpdzApiAggregate'
import { decrypt } from '../crypto/cryptoLib'
import sharesFromTriples from '../math/sharesFromTriples'
import Gfp from '../math/Gfp'

/**
 * Retrieve shares to be used to send input to SPDZ proxies.
 * Wait for all SPDZ proxies to send shares rejecting if an error (no timeout)
 * Validate the triples, sum them
 * @param {inputNum} Number of shares required and so number of triples expected.
 * @param {encrypted} true or false to indicate expect encrypted data
 * @param {spdzProxyList} List containing url and encryptionKey for each proxy 
 * @param {spdzApiRoot} url path
 * @param {clientId} Identify client connection to proxy
 * @returns Promise resolved with list of shares (length inputNum) or reject with Error
 */
const retrieveShares = (inputNum, encrypted, spdzProxyList, spdzApiRoot, clientId) => {
  return consumeDataFromProxies(spdzProxyList.map(spdzProxy => spdzProxy.get('url')), spdzApiRoot, clientId)
    .then((cipherValues) => {
      return (encrypted ? cipherValues.map( (buffer, index) => {
          const encryptionKey = spdzProxyList.get(index).get('encryptionKey')
          return decrypt(encryptionKey, buffer)
          }) : cipherValues)
    })
    .then((clearValues) => {
      try {
        return Promise.resolve(sharesFromTriples(inputNum, clearValues))
      } catch (err) {
        Promise.reject(err)
      }
    })
}

/**
 * Sending a list of inputs to all Spdz proxies after retrieving shares and applying to inputs.
 * @returns Promise with empty return if all OK or rejects with error
 */
const sendInputsWithShares = ( (inputList, encrypted, spdzProxyList, spdzApiRoot, clientId) => {
  return retrieveShares(inputList.length, encrypted, spdzProxyList, spdzApiRoot, clientId )
      .then( (shareList) => {
        return inputList.map( (input, i) => {
          return shareList[i].add(Gfp.fromString(input).toMontgomery())  
        })
      })
      .then( (inputList) => {
        return sendInputsToProxies(spdzProxyList.map(spdzProxy => spdzProxy.get('url')),
                 spdzApiRoot, clientId, inputList)
      })
})

const pollForResult = ( () => {
  // How handle long poll ?
  // See https://davidwalsh.name/javascript-polling for polling with promises, i.e. wait for data.
  // Think about api like readShares, writeIntegers, readIntegers with params (encrypt (true/false), count) )
})

export { retrieveShares, sendInputsWithShares, pollForResult }
