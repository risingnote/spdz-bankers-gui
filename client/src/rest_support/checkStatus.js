/**
 * Provide a function for fetch request to determine if request was successful
 */
export default function checkStatus(response) {
    if (response.ok) {
      return response
    } else {
      var error = new Error(`Status: ${response.status} Message: ${response.statusText}`)
      error.response = response
      throw error
    }
  }