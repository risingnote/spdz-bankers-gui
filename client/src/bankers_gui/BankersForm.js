/**
 * Responsible for managing local state on form before submit.
 */
import React, { Component, PropTypes } from 'react'

import './BankersForm.css'

class BankersForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      bonus: '',
      participantName: ''
    }
    this.handleBonusChange = this.handleBonusChange.bind(this)
    this.handleNameChange = this.handleNameChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleBonusChange(event) {
    this.setState({bonus: event.target.value})
  }

  handleNameChange(event) {
    this.setState({participantName: event.target.value})
  }

  handleSubmit(event) {
    this.props.submitBonus(this.state.participantName, this.state.bonus)
    this.setState({bonus: '', participantName: ''})
    event.preventDefault()
  }

  render() {
    const disableSubmit = this.props.joinedName === undefined && this.props.connectionProblem === undefined ? '' : 'disabled'
    const statusMessage = () => {
      if (this.props.connectionProblem !== undefined) return this.props.connectionProblem
      else if (this.props.winnerChosen) return `A winner has been chosen.`
      else if (this.props.joinedName !== undefined) return `You have joined the meal as ${this.props.joinedName}.`
      else return ''
    }

    return (
        <form className="BankersForm" onSubmit={this.handleSubmit}>
          <h4>Bankers Celebration Dinner</h4>
          <p>...but who should pay?</p>        
          <label htmlFor="joinName">Join meal as</label>
          <input type="text" id="joinName" value={this.state.participantName} onChange={this.handleNameChange} disabled={disableSubmit}/>
          <label htmlFor="bonusValue">Bonus</label>
          <input type="text" id="bonusValue" value={this.state.bonus} onChange={this.handleBonusChange} disabled={disableSubmit}/>
          <input type="submit" value="Send" disabled={disableSubmit}/>
          <p className='smallText'>{statusMessage()}</p>
        </form>
    )
  }
}

BankersForm.propTypes = {
  submitBonus: PropTypes.func.isRequired,
  joinedName: PropTypes.string,
  winnerChosen: PropTypes.bool,
  connectionProblem: PropTypes.string
}

export default BankersForm
