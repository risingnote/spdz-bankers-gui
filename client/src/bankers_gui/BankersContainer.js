/**
 * Responsible for managing data and behaviour for the bankers GUI.
 * Note normally wrapped in SetupContainer to allow interaction with SPDZ proxies.
 * Interacts with GUI via websocket to read and write list of diners who have joined meal.
 * Interacts with SPDZ to send shares of data and poll for result.
 */
import React, { Component, PropTypes } from 'react'
import { List } from 'immutable'
import Io from 'socket.io-client'

import Alert from 'react-s-alert'
import 'react-s-alert/dist/s-alert-default.css'
import 'react-s-alert/dist/s-alert-css-effects/flip.css'

import { connectToProxies, disconnectFromProxies } from '../rest_support/SpdzApiAggregate'
import { sendInputsWithShares, retrieveWinnerClientId } from '../rest_support/SpdzApiHelper'
import NoContentError from '../rest_support/NoContentError'

import BankersForm from './BankersForm'
import BankersTable from './BankersTable'

import './BankersContainer.css'

class BankersContainer extends Component {
  constructor (props) {
    super(props)
    this.state = {
      dinersList: [],
      winningClientId: undefined
    }
    this.socket = undefined
    this.resultTimerId = undefined
    this.handleSubmitEntry = this.handleSubmitEntry.bind(this)
    this.joinMeal = this.joinMeal.bind(this)
    this.pollForResult = this.pollForResult.bind(this)
  }

  /**
   * At startup get list of diners who have already joined the meal.
   */
  componentDidMount() {
    this.socket = Io('/diners')

    // Expect list of {name: <somename>, publicKey: <key>}
    this.socket.on('diners', (msg) => {
      this.setState({dinersList: msg})
    });
  }

  componentWillUnmount() {
    this.socket.on('disconnect', () => {})
    if (this.resultTimerId !== undefined) {
      clearInterval(this.resultTimerId)
      this.resultTimerId = undefined
    } 
    disconnectFromProxies(this.props.spdzProxyServerList.map( spdzProxy => spdzProxy.get('url')), 
                          this.props.spdzApiRoot, this.props.clientPublicKey)
  }

  /**
   * Start an interval timer to poll the SPDZ proxies for the winner of the computation.
   * If successfully get result:
   *   Disconnect from SPDZ engines and notify ConnectionContainer.
   */
  pollForResult(spdzProxyServerList, spdzApiRoot, clientPublicKey) {
    this.resultTimerId = setInterval(() => {
      retrieveWinnerClientId(spdzProxyServerList, spdzApiRoot, clientPublicKey)
      .then( winningClientId => {
        this.setState({winningClientId: winningClientId})
        console.log('Got winning client of ', winningClientId)        
        clearInterval(this.resultTimerId)
      })
      .then( () => {
        return disconnectFromProxies(this.props.spdzProxyServerList.map( spdzProxy => spdzProxy.get('url')), 
                          this.props.spdzApiRoot, this.props.clientPublicKey)
      })
      .then( (values) => {
        this.props.stopGame(values)
      })
      .catch( err => {
        if (err instanceof NoContentError) {
          // Keep going, data should turn up.
          console.log('No result data yet.')
        }
        else {
          console.log('Problem when polling for result ', err)
          clearInterval(this.resultTimerId)        
        }
      })
    }, 2000)
  }

  /**
   * Join meal by sending websocket message, wrapped as promise.
   */
  joinMeal(socket, name, clientPublicKey) {
    return new Promise(function(resolve, reject) {
      socket.emit('joinMeal', {name: name, publicKey: clientPublicKey}, function(error) {
          if (error) {
            reject(error)
          } else {
            resolve()
          }
      })
    })
  } 

  /**
   * Join game by:
   *   Running connection to SPDZ engines and notifying ConnectionContainer.
   *   If all OK join meal with a name
   *   If all OK send bonus as shares to all SPDZ proxies.
   *   Start timer polling for result.
   */
  handleSubmitEntry(name, bonus) {
    connectToProxies(this.props.spdzProxyServerList.map( spdzProxy => spdzProxy.get('url')), 
                                this.props.spdzApiRoot, this.props.clientPublicKey)
      .then( (values) => {
        this.props.startGame(values)
        return this.joinMeal(this.socket, name, this.props.clientPublicKey)
      })
      .then( () => {
        return sendInputsWithShares([bonus], true, this.props.spdzProxyServerList, 
              this.props.spdzApiRoot, this.props.clientPublicKey)
      })
      .then( () => {
        this.pollForResult(this.props.spdzProxyServerList, this.props.spdzApiRoot, 
              this.props.clientPublicKey)
        Alert.info('You have joined the meal.', {html: false})
      })
      .catch((ex) => {
        console.log(ex)
        Alert.error(`<h4>Unable to successfully send bonus to SPDZ proxies.</h4><p>${ex}</p>`, {timeout: 'none'})
      })
  }

  render() {
    const myEntry = this.state.dinersList.find(diner => diner.publicKey === this.props.clientPublicKey)
    const joinedName = (myEntry !== undefined ? myEntry.name : undefined)
    const winnerChosen = this.state.winningClientId !== undefined
    const connectionProblem = this.props.spdzProxyServerList.size === 0 ? "No SPDZ servers found." : undefined

    return (
      <div className='Bankers'>
        <BankersForm submitBonus={this.handleSubmitEntry} joinedName={joinedName} 
                     winnerChosen={winnerChosen} connectionProblem={connectionProblem}/>
        <BankersTable diners={this.state.dinersList} winningClientId={this.state.winningClientId}/>
        <Alert stack={{limit: 3}} timeout={5000} position={'top-left'} effect={'flip'} offset={100} html={true} />
      </div>
    )
  }
}

BankersContainer.propTypes = {
  startGame: PropTypes.func.isRequired,
  stopGame: PropTypes.func.isRequired,  
  spdzProxyServerList: PropTypes.instanceOf(List).isRequired,
  spdzApiRoot: PropTypes.string.isRequired,
  clientPublicKey: PropTypes.string.isRequired
}

export default BankersContainer
