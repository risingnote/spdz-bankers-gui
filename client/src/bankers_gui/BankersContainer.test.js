import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import BankersContainer from './BankersContainer'
import BankersForm from './BankersForm'
import { twoProxiesWith1Connected, twoProxiesWith2Connected } from '../test_support/ProxyServerList'

// Mock out REST call
jest.mock('../rest_support/SpdzApiHelper')
import { sendInputsWithShares } from '../rest_support/SpdzApiHelper'

describe('Bankers GUI (not wrapped in SetupContainer) rendering and behaviour', () => {
  it('Renders as expected (compared to a snapshot) when proxies are not all connected', () => {
    const tree = renderer.create(
      <BankersContainer allProxiesConnected={false} spdzProxyServerList={twoProxiesWith1Connected}
                        spdzApiRoot={'/spdzapi'} clientPublicKey={'a1b2c3d4'} />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Renders as expected (compared to a snapshot) when proxies are connected', () => {
    const tree = renderer.create(
      <BankersContainer allProxiesConnected={true} spdzProxyServerList={twoProxiesWith2Connected}
                        spdzApiRoot={'/spdzapi'} clientPublicKey={'a1b2c3d4'} />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('Handles the submission of the bonus', () => {
    sendInputsWithShares.mockImplementation(() => Promise.resolve())

    const wrapper = mount(
      <BankersContainer allProxiesConnected={true} spdzProxyServerList={twoProxiesWith2Connected}
                        spdzApiRoot={'/spdzapi'} clientPublicKey={'a1b2c3d4'} />
    )

    const form = wrapper.find(BankersForm)
    const input = form.find('[id="bonusValue"]') //property selector find
    //Simulate user input followed by click submit
    const userInput = 123
    input.simulate('change', {target: {value: userInput}})
    form.simulate('submit')

    expect(sendInputsWithShares.mock.calls.length).toEqual(1)
    expect(sendInputsWithShares.mock.calls[0]).toEqual([[userInput], true, twoProxiesWith2Connected, '/spdzapi', 'a1b2c3d4'])

    sendInputsWithShares.mockClear()
  })
})
