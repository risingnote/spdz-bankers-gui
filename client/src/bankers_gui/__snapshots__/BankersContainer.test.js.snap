exports[`Bankers GUI (not wrapped in SetupContainer) rendering and behaviour Renders as expected (compared to a snapshot) at start of game 1`] = `
<div
  className="Bankers">
  <form
    className="BankersForm"
    onSubmit={[Function]}>
    <h4>
      Bankers Celebration Dinner
    </h4>
    <p>
      ...but who should pay?
    </p>
    <label
      htmlFor="joinName">
      Join meal as
    </label>
    <input
      disabled=""
      id="joinName"
      onChange={[Function]}
      type="text"
      value="" />
    <label
      htmlFor="bonusValue">
      Bonus
    </label>
    <input
      disabled=""
      id="bonusValue"
      onChange={[Function]}
      type="text"
      value="" />
    <input
      disabled=""
      type="submit"
      value="Send" />
    <p
      className="smallText">
      
    </p>
  </form>
  <div
    className="BankersTable">
    <div
      className="place"
      style={
        Object {
          "transform": "translate(0rem, -8.2rem) rotate(180deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-180deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(5.8rem, -5.8rem) rotate(225deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-225deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(8.2rem, 0rem) rotate(270deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-270deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(5.8rem, 5.8rem) rotate(315deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-315deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(0rem, 8.2rem) rotate(0deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-0deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(-5.8rem, 5.8rem) rotate(45deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-45deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(-8.2rem, 0rem) rotate(90deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-90deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(-5.8rem, -5.8rem) rotate(135deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-135deg)",
            }
          }>
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="s-alert-wrapper" />
</div>
`;

exports[`Bankers GUI (not wrapped in SetupContainer) rendering and behaviour Renders as expected (compared to a snapshot) at start of game, no proxies found 1`] = `
<div
  className="Bankers">
  <form
    className="BankersForm"
    onSubmit={[Function]}>
    <h4>
      Bankers Celebration Dinner
    </h4>
    <p>
      ...but who should pay?
    </p>
    <label
      htmlFor="joinName">
      Join meal as
    </label>
    <input
      disabled="disabled"
      id="joinName"
      onChange={[Function]}
      type="text"
      value="" />
    <label
      htmlFor="bonusValue">
      Bonus
    </label>
    <input
      disabled="disabled"
      id="bonusValue"
      onChange={[Function]}
      type="text"
      value="" />
    <input
      disabled="disabled"
      type="submit"
      value="Send" />
    <p
      className="smallText">
      No SPDZ servers found.
    </p>
  </form>
  <div
    className="BankersTable">
    <div
      className="place"
      style={
        Object {
          "transform": "translate(0rem, -8.2rem) rotate(180deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-180deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(5.8rem, -5.8rem) rotate(225deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-225deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(8.2rem, 0rem) rotate(270deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-270deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(5.8rem, 5.8rem) rotate(315deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-315deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(0rem, 8.2rem) rotate(0deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-0deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(-5.8rem, 5.8rem) rotate(45deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-45deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(-8.2rem, 0rem) rotate(90deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-90deg)",
            }
          }>
          
        </div>
      </div>
    </div>
    <div
      className="place"
      style={
        Object {
          "transform": "translate(-5.8rem, -5.8rem) rotate(135deg)",
        }
      }>
      <div
        className="place-setting">
        <div
          style={
            Object {
              "transform": "rotate(-135deg)",
            }
          }>
          
        </div>
      </div>
    </div>
  </div>
  <div
    className="s-alert-wrapper" />
</div>
`;
